i2c/imx6ull/imx6ull-i2c.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Phoenix-RTOS
+ *
+ * i.MX 6ULL I2C BUS driver
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Marek Bialowas
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <sys/msg.h>
+#include <posix/utils.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <i2c.h>
+#include <i2c-msg.h>
+
+
+static int dev_ctl(msg_t *msg)
+{
+	i2c_devctl_t *i = (i2c_devctl_t *)msg->i.raw;
+
+	switch (i->type) {
+		case i2c_devctl_bus_write:
+			return i2c_busWrite(i->dev_addr, msg->i.data, msg->i.size);
+
+		case i2c_devctl_bus_read:
+			return i2c_busRead(i->dev_addr, msg->o.data, msg->o.size);
+
+		case i2c_devctl_reg_read:
+			return i2c_regRead(i->dev_addr, i->reg_addr, msg->o.data, msg->o.size);
+
+		default:
+			return -EINVAL;
+	}
+}
+
+
+static void thread(void *arg)
+{
+	uint32_t port = (uint32_t)arg;
+	msg_t msg;
+	unsigned long int rid;
+
+	for (;;) {
+		if (msgRecv(port, &msg, &rid) < 0) {
+			memset(&msg, 0, sizeof(msg));
+			msgRespond(port, &msg, rid);
+			continue;
+		}
+
+		switch (msg.type) {
+			case mtOpen:
+			case mtClose:
+				msg.o.io.err = EOK;
+				break;
+			case mtRead:
+			case mtWrite:
+				msg.o.io.err = -EINVAL;
+				break;
+			case mtDevCtl:
+				msg.o.io.err = dev_ctl(&msg);
+				break;
+			default:
+				msg.o.io.err = -EINVAL;
+				break;
+		}
+
+		msgRespond(port, &msg, rid);
+	}
+}
+
+
+static void print_usage(const char *progname)
+{
+	printf("Usage: %s [i2c_bus_no <1,4>]\n", progname);
+}
+
+
+int main(int argc, char **argv)
+{
+	oid_t dev;
+	unsigned int dev_no;
+	uint32_t port;
+	char devname[sizeof("i2cX")];
+
+	if (argc != 2) {
+		print_usage(argv[0]);
+		return 1;
+	}
+
+	dev_no = atoi(argv[1]);
+	if (i2c_init(dev_no) < 0) {
+		printf("i2c: initialization failed\n");
```

Comment:
It is better to send error messages to `stderr`.

Reply:
For userspace programs - yes, for drivers - not really. Right now it's messy (stdout/stderr pinned to console so no possibility to read the output from drivers remotely).

The obvious choice for now would be to use syslog() to log all of the driver messages, but we decided to implement dmesg-alike circular buffer (with `/dev/kmsg` interface) and pin it to stdout of all of the drivers and than use syslogd to automatically pump the logs from `/dev/kmsg` to syslog (on bigger systems) or be able do inspect dmesg buffer directly.  **Please note that using line-buffered stream (`stdout` as opposed to `stderr`) would be much more useful then to prevent mixing log messages from multiple processes**.

I don't see any real benefit for using `stderr` now, I would much more welcome dedicated `log` wrappers which will add message priority to the dmesg messages (along with the introduction of `dmesg`).

Leaving as-is on purpose.